### PR TITLE
Publish setting changes

### PR DESCRIPTION
**Pull request by bartoszWojciechO**

### CODE REVIEW

File: protobuf/state/state.proto [hunk 1]
```diff
@@ -12,6 +15,44 @@ enum ConnectionState {
...
+	THREATPROTECTIONLITE = 1;
+	OBFUSCATE = 2;
+	FIREWALL = 3;
+	ROUTING = 4;
+	NOTIFY = 5;
+	MESHNET = 6;
+	IPV6 = 7;
+	DEFAULTS = 8;
+	LAN_DISCOVERY = 9;
+	VIRTUA_LOCATION = 10;
```
Comment (keliramu) [inline protobuf/state/state.proto:29]: Typo `VIRTUA_LOCATION`

File: state/state.go [hunk 2]
```diff
@@ -44,11 +50,17 @@ func (s *StatePublisher) notify(e interface{}) {
...
+	defer s.mu.Unlock()
+	s.notify(SettingEvent{setting: settingType, enabled: enabled})
+}
+
 func (s *StatePublisher) NotifyConnect(e events.DataConnect) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	log.Printf(internal.DebugPrefix+" notifying about connect event: %+v", e)
+	log.Println(internal.DebugPrefix, "notifying about connect event: %+v", e)
```
Comment (devzbysiu) [inline state/state.go:63]: Will this formatting string here work with `log.Println`?

File: protobuf/state/state.proto [hunk 3]
```diff
@@ -12,6 +15,44 @@ enum ConnectionState {
     CONNECTED = 2;
 }
 
+enum Setting {
+    AUTOCONNECT = 0;
+	THREATPROTECTIONLITE = 1;
```
Comment (devzbysiu) [inline protobuf/state/state.proto:20]: Formatting? Or is this indentation intentional? :smile:

**Review by devzbysiu — APPROVED** [review]

File: protobuf/state/state.proto [hunk 4]
```diff
@@ -22,8 +63,23 @@ message ConnectionStatus {
...
+}
+
+message LoginEvent {
+    LoginEventType type = 1;
+}
+
 message AppState {
     oneof state {
         ConnectionStatus connection_status = 1;
+        SettingChange setting_change = 2;
```
Comment (keliramu) [inline protobuf/state/state.proto:78]: Meshnet state to be added later?

Comment (bartoszWojciechO) [reply protobuf/state/state.proto:78]: For this task there's a notification for meshnet setting change(on/off). I don't see any task for more complicated meshnet data, such as new peers appearing or changing permissions, maybe it should be added. :thinking:

File: config/manager.go [hunk 5]
```diff
@@ -85,19 +89,22 @@ type FilesystemConfigManager struct {
...
+	configPublisher ConfigPublisher
 	mu              sync.Mutex
 }
 
 // NewFilesystemConfigManager is constructed from a given location and salt.
 func NewFilesystemConfigManager(location, vault, salt string,
 	machineIDGetter MachineIDGetter,
-	fsHandle FilesystemHandle) *FilesystemConfigManager {
+	fsHandle FilesystemHandle,
+	conifgPublisher ConfigPublisher) *FilesystemConfigManager {
```
Comment (mariusSincovici) [inline config/manager.go:100]: [typo] `conifgPublisher`

File: config/manager.go [hunk 6]
```diff
@@ -114,7 +121,11 @@ func (f *FilesystemConfigManager) SaveWith(fn SaveFunc) error {
 	}
 
 	c = fn(c)
-	return f.save(c)
+	err := f.save(c)
+	if err == nil && f.configPublisher != nil {
+		f.configPublisher.Publish(&c)
```
Comment (mariusSincovici) [inline config/manager.go:126]: Shouldn't this not be called while the mutex is locked?

Comment (bartoszWojciechO) [reply config/manager.go:126]: Why?

**Review by mariusSincovici — APPROVED** [review]

File: config/manager.go (hunk 6 above)
Comment (mariusSincovici) [reply config/manager.go:126]: The only concern is not to have a deadlock,  if some component will listening on this events and then tries to read the settings from here.

**Pull request merged**